### DISCUSSION TITLE
RFC3161 support ?

### DISCUSSION THREAD

**Discussion by dgillier** [lsnepomuceno/laravel-a1-pdf-sign#109] [General]
Hello !
Any chance to get RFC3161 time stamping support ?

Thanks for considering.
Denis

**Comment by lsnepomuceno** [comment]
Hello @dgillier,

Yes — RFC 3161 timestamping shipped in **2.0**.

```php
use LSNepomuceno\LaravelA1PdfSign\Facades\A1PdfSign;

$signed = A1PdfSign::newSignature()
    ->certificate($pfxPath, $password)
    ->pdf($pdfPath)
    ->timestamp()          // shorthand for the pades-b-t profile
    ->sign();
```

The authority comes from configuration — `A1_TSA_URL`, with optional `A1_TSA_USERNAME` / `A1_TSA_PASSWORD` — so it is not hard-coded at the call site. freetsa.org works; the test suite has a group that signs against it live.

Timestamping is one level of a profile ladder rather than a standalone flag, because the levels above it need it:

| Profile | Adds |
|---|---|
| `pades-b-b` | CAdES signed attributes. The default |
| `pades-b-t` | **an RFC 3161 timestamp** |
| `pades-b-lt` | a Document Security Store — embedded OCSP responses and CRLs, so the signature still verifies after the certificate expires |
| `pades-b-lta` | an archive timestamp over the whole file |

```php
->profile('pades-b-lta')
```

Signatures are PAdES baseline by default and carry the ESS `signing-certificate-v2` attribute, which `openssl_pkcs7_sign()` cannot emit — that is why the CMS is built with `tc-lib-pdf-sign` rather than through OpenSSL directly.

Signed samples for every profile are in [`samples/`](https://github.com/lsnepomuceno/laravel-a1-pdf-sign/tree/main/samples), and all of them were verified independently with poppler's `pdfsig`.

Thanks for asking back then, and sorry for the long delay.
